File: tools/covenant/critic.py
#!/usr/bin/env python3
"""Critic module for Covenant orchestration v2."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

ALLOWED_AGENTS = {
    "agent.monitor.v1",
    "agent.huragok.v1",
    "agent.oracle.v1",
    "agent.librarian.v1",
}

DEFAULT_BUDGET = {
    "max_total_timeout_seconds": 7200,
    "max_total_retries": 8,
}


def review_plan(plan: dict[str, Any], request: dict[str, Any] | None = None) -> dict[str, Any]:
    request = request or {}
    issues: list[str] = []
    warnings: list[str] = []

    steps = plan.get("steps")
    if not isinstance(steps, list) or not steps:
        return {
            "approved": False,
            "requires_human_review": True,
            "issues": ["plan.steps must be a non-empty array"],
            "warnings": [],
            "resource_budget": {"total_timeout_seconds": 0, "total_retries": 0},
        }

    step_ids = {s.get("step_id") for s in steps if isinstance(s, dict)}
    total_timeout = 0
    total_retries = 0

    for idx, step in enumerate(steps):
        if not isinstance(step, dict):
            issues.append(f"steps[{idx}] must be an object")
            continue

        agent = step.get("agent_identity_id")
        if agent not in ALLOWED_AGENTS:
            issues.append(f"steps[{idx}] has unknown agent_identity_id '{agent}'")

        conf = step.get("confidence")
        threshold = step.get("confidence_threshold")
        if not isinstance(conf, (int, float)) or not isinstance(threshold, (int, float)):
            issues.append(f"steps[{idx}] confidence/threshold must be numeric")
        elif conf < threshold:
            issues.append(
                f"steps[{idx}] confidence {conf:.2f} below threshold {threshold:.2f}; requires re-planning or human review"
            )

        deps = step.get("depends_on", [])
        if not isinstance(deps, list):
            issues.append(f"steps[{idx}] depends_on must be an array")
        else:
            for dep in deps:
                if dep not in step_ids:
                    issues.append(f"steps[{idx}] depends_on unknown step_id '{dep}'")

        retry = step.get("retry_policy", {})
        if not isinstance(retry, dict):
            issues.append(f"steps[{idx}] retry_policy must be an object")
            retry = {}

        timeout = retry.get("timeout_seconds", 0)
        retries = retry.get("max_retries", 0)
        if isinstance(timeout, int) and timeout >= 0:
            total_timeout += timeout
        else:
            issues.append(f"steps[{idx}] retry_policy.timeout_seconds must be integer >= 0")
        if isinstance(retries, int) and retries >= 0:
            total_retries += retries
        else:
            issues.append(f"steps[{idx}] retry_policy.max_retries must be integer >= 0")

        gate = step.get("quality_gate", {})
        if not isinstance(gate, dict) or not gate.get("checks"):
            issues.append(f"steps[{idx}] must define a quality_gate with checks")

        handoff = step.get("handoff", {})
        if not isinstance(handoff, dict) or "output_contract" not in handoff:
            issues.append(f"steps[{idx}] must define handoff contract")

    budget_req = request.get("resource_budget") if isinstance(request.get("resource_budget"), dict) else {}
    max_timeout = int(budget_req.get("max_total_timeout_seconds", DEFAULT_BUDGET["max_total_timeout_seconds"]))
    max_retries = int(budget_req.get("max_total_retries", DEFAULT_BUDGET["max_total_retries"]))

    if total_timeout > max_timeout:
        issues.append(f"timeout budget exceeded: {total_timeout}s > {max_timeout}s")
    if total_retries > max_retries:
        issues.append(f"retry budget exceeded: {total_retries} > {max_retries}")

    if len(steps) > int(request.get("max_steps", 6)):
        warnings.append(f"plan uses {len(steps)} steps; exceeds preferred max_steps")

    approved = not issues
    requires_human = any("human review" in i.lower() for i in issues) or not approved

    return {
        "approved": approved,
        "requires_human_review": requires_human,
        "issues": issues,
        "warnings": warnings,
        "resource_budget": {
            "total_timeout_seconds": total_timeout,
            "total_retries": total_retries,
        },
    }


def main() -> None:
    import argparse

    parser = argparse.ArgumentParser(description="Critic review for Covenant plan")
    parser.add_argument("plan", help="Path to plan JSON")
    parser.add_argument("--request", help="Path to request JSON", default=None)
    args = parser.parse_args()

    plan = json.loads(Path(args.plan).expanduser().resolve().read_text())
    request = json.loads(Path(args.request).expanduser().resolve().read_text()) if args.request else {}
    print(json.dumps(review_plan(plan, request), indent=2))


if __name__ == "__main__":
    main()

File: tools/covenant/critic.ts
#!/usr/bin/env npx tsx

/** Arbiter module for Covenant orchestration v2. */

import fs from "fs";
import path from "path";

const ALLOWED_AGENTS = new Set([
  "agent.monitor.v1",
  "agent.huragok.v1",
  "agent.researcher.v1",
  "agent.oracle.v1",
  "agent.librarian.v1",
]);

const DEFAULT_BUDGET = {
  max_total_timeout_seconds: 7200,
  max_total_retries: 8,
};

type Json = Record<string, unknown>;

type ReviewResult = {
  approved: boolean;
  requires_human_review: boolean;
  issues: string[];
  warnings: string[];
  resource_budget: { total_timeout_seconds: number; total_retries: number };
};

export function reviewPlan(plan: Json, request: Json | null = null): ReviewResult {
  const req = request ?? {};
  const issues: string[] = [];
  const warnings: string[] = [];

  const steps = plan.steps;
  if (!Array.isArray(steps) || steps.length === 0) {
    return {
      approved: false,
      requires_human_review: true,
      issues: ["plan.steps must be a non-empty array"],
      warnings: [],
      resource_budget: { total_timeout_seconds: 0, total_retries: 0 },
    };
  }

  const stepIds = new Set(
    steps
      .filter((s) => s && typeof s === "object")
      .map((s) => (s as Json).step_id)
      .filter((id) => typeof id === "string") as string[]
  );

  let totalTimeout = 0;
  let totalRetries = 0;

  steps.forEach((step, idx) => {
    if (!step || typeof step !== "object") {
      issues.push(`steps[${idx}] must be an object`);
      return;
    }
    const stepObj = step as Json;

    const agent = stepObj.agent_identity_id;
    if (typeof agent !== "string" || !ALLOWED_AGENTS.has(agent)) {
      issues.push(`steps[${idx}] has unknown agent_identity_id '${String(agent)}'`);
    }

    const conf = stepObj.confidence;
    const threshold = stepObj.confidence_threshold;
    if (typeof conf !== "number" || typeof threshold !== "number") {
      issues.push(`steps[${idx}] confidence/threshold must be numeric`);
    } else if (conf < threshold) {
      issues.push(
        `steps[${idx}] confidence ${conf.toFixed(2)} below threshold ${threshold.toFixed(2)}; requires re-planning or human review`
      );
    }

    const deps = stepObj.depends_on;
    if (!Array.isArray(deps)) {
      issues.push(`steps[${idx}] depends_on must be an array`);
    } else {
      for (const dep of deps) {
        if (typeof dep !== "string" || !stepIds.has(dep)) {
          issues.push(`steps[${idx}] depends_on unknown step_id '${String(dep)}'`);
        }
      }
    }

    const retry = stepObj.retry_policy;
    const retryObj = retry && typeof retry === "object" ? (retry as Json) : null;
    if (!retryObj) {
      issues.push(`steps[${idx}] retry_policy must be an object`);
    }

    const timeout = retryObj ? retryObj.timeout_seconds : null;
    const retries = retryObj ? retryObj.max_retries : null;

    if (typeof timeout === "number" && Number.isInteger(timeout) && timeout >= 0) {
      totalTimeout += timeout;
    } else {
      issues.push(`steps[${idx}] retry_policy.timeout_seconds must be integer >= 0`);
    }

    if (typeof retries === "number" && Number.isInteger(retries) && retries >= 0) {
      totalRetries += retries;
    } else {
      issues.push(`steps[${idx}] retry_policy.max_retries must be integer >= 0`);
    }

    const gate = stepObj.quality_gate;
    if (!gate || typeof gate !== "object" || !(gate as Json).checks) {
      issues.push(`steps[${idx}] must define a quality_gate with checks`);
    }

    const handoff = stepObj.handoff;
    if (!handoff || typeof handoff !== "object" || !(handoff as Json).output_contract) {
      issues.push(`steps[${idx}] must define handoff contract`);
    }
  });

  const budgetReq = req.resource_budget && typeof req.resource_budget === "object" ? (req.resource_budget as Json) : {};
  const maxTimeout = Number.parseInt(String(budgetReq.max_total_timeout_seconds ?? DEFAULT_BUDGET.max_total_timeout_seconds), 10);
  const maxRetries = Number.parseInt(String(budgetReq.max_total_retries ?? DEFAULT_BUDGET.max_total_retries), 10);

  if (totalTimeout > maxTimeout) {
    issues.push(`timeout budget exceeded: ${totalTimeout}s > ${maxTimeout}s`);
  }
  if (totalRetries > maxRetries) {
    issues.push(`retry budget exceeded: ${totalRetries} > ${maxRetries}`);
  }

  const maxSteps = Number.parseInt(String(req.max_steps ?? 6), 10);
  if (steps.length > maxSteps) {
    warnings.push(`plan uses ${steps.length} steps; exceeds preferred max_steps`);
  }

  const approved = issues.length === 0;
  const requiresHuman = issues.some((i) => i.toLowerCase().includes("human review")) || !approved;

  return {
    approved,
    requires_human_review: requiresHuman,
    issues,
    warnings,
    resource_budget: { total_timeout_seconds: totalTimeout, total_retries: totalRetries },
  };
}

async function main(): Promise<void> {
  const args = process.argv.slice(2);
  if (args.length < 1) {
    console.error("Usage: critic.ts <plan.json> [--request <request.json>]");
    process.exit(2);
  }

  const planPath = path.resolve(args[0]);
  let requestPath: string | null = null;
  const requestIndex = args.indexOf("--request");
  if (requestIndex !== -1 && args[requestIndex + 1]) {
    requestPath = path.resolve(args[requestIndex + 1]);
  }

  const plan = JSON.parse(fs.readFileSync(planPath, "utf8"));
  const request = requestPath ? JSON.parse(fs.readFileSync(requestPath, "utf8")) : {};
  const result = reviewPlan(plan, request);
  console.log(JSON.stringify(result, null, 2));
}

main();
